README
This is GLOST, the Greedy Launcher Of Small Tasks. 

== Requirements ==

MPI

== Compil & Install ==

edit the Makefile.
$ make

== Example ==

see the scripts in directory examples/

== Documentation ==

* see ./glost_launch -h
* see ./glost_filter -h
* see ./glost_bcast

== Other ==

If you want a new feature, please contact:

bruno.froge@cea.fr

In case of a bug, please contact us at:

vincent.ducrot.tgcc@cea.fr
yohan.lee-tin-yien.tgcc@cea.fr



NEWS
- add option "--maxline_length" and its example
- security fixe(s)
- add example for glost_bcast

== since 0.3.0 ==

- add glost_bcast
- fix segfault on no input file
- fix compilation on titane
- other code quality fixes

== since 0.2.4 ==

- fix the examples (due to the software rename)
- add "-h" option in glost_launch

== since 0.2.0 ==

- rename the soft to "glost"
- add feature: support of -R <time_remaining> option
- add feature: support of signal SIGUSR1 to properly stop the glost_launch
- Bugfix : output status matches the exit code of launched command
- Bugfix : empty line in input file kills worker.
- add example : example/test_resume.* , showing how a ccc_msub job which can resume
- enhance the basic example and add new ones for each new feature.

== since 0.1 ==
- initial working version


examples/test_launch.sh
#!/bin/bash
mpirun -np 9 ../glost_launch test_launch.list


examples/test_resume.sh
#!/bin/bash

for i in $(seq 1 3); do
    echo "launch number $i"
    jobid=`ccc_msub test_resume.msub | cut -d ' ' -f4` # launching
    sleep 30s                                            # waiting
    ccc_mdel -s SIGUSR1 $jobid                           # killing properly
done

examples/test_sigusr1.sh
#!/bin/bash

../glost_launch test_sigusr1.list &
pid=$!
sleep 5
kill -s SIGUSR1 $pid


glost_bcast.c
/*
   GLOST - Greedy Launcher Of Small Tasks.
   Copyright (C) 2014  Bruno FROGE <bruno.froge@cea.fr>

   Author: Vincent DUCROT     <vincent.ducrot.tgcc@cea.fr>
   Author: Yohan LEE-TIN-YIEN <yohan.lee-tin-yien.tgcc@cea.fr>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA

*/

/* Usage:
 *    glost_bcast FILE[...]
 *
 * Broadcast the file FILE into TMPDIR
 *
 *
 * Author: Yohan LEE-TIN-YIEN <yohan.lee-tin-yien.tgcc@cea.fr>
 */

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <libgen.h>
#include <assert.h>


#ifndef  DEFAULT_BUFFLEN
#define DEFAULT_BUFFLEN 134217728
#endif

#define MPI_INT_ISTRUE  1
#define MPI_INT_ISFALSE 0

#define VERSION_MAJOR   0
#define VERSION_MINOR   3
#define VERSION_PATCH   1

#define MAX(x, y) (((x) > (y)) ? (x) : (y))
#define MIN(x, y) (((x) < (y)) ? (x) : (y))

/* http://programanddesign.com/cpp/human-readable-file-size-in-c/ */
char* readable_fs(double size/*in bytes*/, char *buf) {
  int i = 0;
  const char* units[] = {"B", "kB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"};
  while (size > 1024) {
    size /= 1024;
    i++;
      }
  sprintf(buf, "%.*f %s", i, size, units[i]);
  return buf;
}

void usage(int argc,char **argv){
  printf("usage: [mpirun] [mpirun_options] %s FILE [...]\n\n"
	 "%s copies a shared FILE into each local directory $TMPDIR using MPI-IO.\n"
	 "Copies are done by creating the temporary files $TMPDIR/FILE.<rank> and a final renaming.\n\n"
	 "Version: \n  %i.%i.%i \n\n"
	 "Examples:\n"
	 "  mpirun -n 4 %s file \n"
	 "  ccc_mprun -n 4 -c 16 %s file \n"
	 ,basename(argv[0]),basename(argv[0])
	 ,VERSION_MAJOR,VERSION_MINOR,VERSION_PATCH
	 ,argv[0],argv[0]
	 );
}


/* ******************************* */

int main(int argc,char **argv)
{
  int rank,size;
  double timing;

  /* init */
  MPI_Init(&argc,&argv);

  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  MPI_Comm_size(MPI_COMM_WORLD,&size);

  /* read */
  const int filenamesz=4095;
  char *odirname, *ibname;
  char ifilename[filenamesz+1];
  char ofilename[filenamesz+1];
  char ofilename2[filenamesz+1];
  MPI_File ifh,ofh;
  MPI_Info ifinfo;
  char ctmp[100];
  char ctmp2[100];
  char ctmp3[100];

  size_t  bufflen = DEFAULT_BUFFLEN ;
  int iamode,oamode;

  long long fsize, bwritten, bread ;
  long long count;

  char iinfokey[MPI_MAX_INFO_KEY];
  int  iflag;
  int  ierr;

  /* early exit */
  if ((argc == 1) && (rank == 0))
      usage(argc,argv);

  sprintf(ofilename2,"%s","unset");

  for (int arg=1; arg<argc; arg++){
    char *buff1,*buff2;
    /*
     * modes
     */

    iamode = (int) 0 ;
    iamode |= MPI_MODE_RDONLY ;
    iamode |= MPI_MODE_UNIQUE_OPEN ;

    oamode = (int) 0 ;
    oamode |= MPI_MODE_CREATE ;
    oamode |= MPI_MODE_WRONLY ;
    oamode |= MPI_MODE_UNIQUE_OPEN ;

    /*
     * Infos
     */

    MPI_Info_create(&ifinfo);
    MPI_Info_set(ifinfo,"access_style","read_once");
    MPI_Info_set(ifinfo,"collective_buffering","true");
    /* MPI_Info_set(ifinfo,"cb_block_size","1048576"); */ /* 1 MBytes - should match FS block size */
    MPI_Info_set(ifinfo,"cb_block_size","4194304"); /* 4 MBytes - should match FS block size */
    MPI_Info_set(ifinfo,"cb_buffer_size","134217728"); /* 128 MBytes (Optional) */

    /*
     *  filenames
     */

    /* sprintf(ifilename,"%s","./testin.bin"); */
    snprintf(ifilename,filenamesz,"%s",argv[arg]);
    ibname = basename(ifilename);
    /* odirname = secure_getenv("TMPDIR"); */
    odirname = getenv("TMPDIR");
    if (odirname == NULL){
      fprintf(stderr,"#Error: TMPDIR environment variable must be set");
      MPI_Abort(MPI_COMM_WORLD,1);
    }
    odirname = strdup(odirname);
    snprintf(ofilename,filenamesz,"%s/%s.%i",odirname,ibname,rank);
    snprintf(ofilename2,filenamesz,"%s/%s",odirname,ibname);
    free(odirname);
    timing = MPI_Wtime();

    /*
     * Open
     */

    ierr = MPI_File_open(MPI_COMM_WORLD,ifilename, iamode, ifinfo, &ifh);
    if( ierr != MPI_SUCCESS ){
      if (rank == 0 )
	fprintf(stderr,"#Warning: cannot open %s, ignoring it\n", ifilename );
      continue;
    }

#ifndef DO_NOT_WRITE
    ierr = MPI_File_open(MPI_COMM_SELF,ofilename, oamode, MPI_INFO_NULL,&ofh);
    assert( ierr == MPI_SUCCESS );
#endif

    /* match the cb_nodes and striping */
    MPI_File_get_info(ifh,&ifinfo);
    MPI_Info_get(ifinfo,"striping_factor",MPI_MAX_INFO_KEY,iinfokey,&iflag);
    if( iflag == MPI_INT_ISTRUE ){
      if ( rank == 0)
	fprintf(stderr,"#stripping detected, setting cb_nodes : %s .\n", iinfokey );
      MPI_Info_set(ifinfo,"cb_nodes",iinfokey);

      /* reopen to be sure */
      ierr = MPI_File_close(&ifh);
      assert( ierr == MPI_SUCCESS );

      ierr = MPI_File_open(MPI_COMM_WORLD,ifilename, iamode, ifinfo, &ifh);
      assert( ierr == MPI_SUCCESS );
    }

    /*
     * Transfert
     */

    MPI_File_get_size(ifh, &fsize);
#ifndef DO_NOT_WRITE
    MPI_File_preallocate(ofh, fsize);
#endif
    buff1 = malloc(bufflen);
    buff2 = malloc(bufflen);

    if (rank == 0)
      fprintf(stderr,"Each process uses 2 buffers of %s \n",
	      readable_fs( (double) bufflen, ctmp3) );

    /* Assume read bandwidth >> write bandwidth
     * Thus non-blocking read minimise waiting time
     */

    bwritten = 0;
    bread    = 0;
    count    = 0;
    size_t  bufflen1 = bufflen ;
    size_t  bufflen2 = bufflen ;

    /* Handle first read */
    bufflen1 = MIN(fsize-bwritten,bufflen1) ;
    ierr = MPI_File_read_at_all_begin(ifh, bread, buff1, bufflen1, MPI_BYTE);
    assert( ierr == MPI_SUCCESS );
    /*  */
    while( bwritten < fsize){
      if (count%2 == 0 ){
	ierr = MPI_File_read_at_all_end(ifh,buff1,MPI_STATUS_IGNORE);
	assert( ierr == MPI_SUCCESS );
	bufflen1 = MIN(fsize-bwritten,bufflen1) ;
	bread    += bufflen1 ;

	bufflen2 = MIN(fsize-bread   ,bufflen2);
	if ( bufflen2 > 0 )
	  ierr = MPI_File_read_at_all_begin(ifh, bread, buff2, bufflen2, MPI_BYTE);
	assert( ierr == MPI_SUCCESS );

#ifndef DO_NOT_WRITE
	ierr = MPI_File_write_at(ofh, bwritten, buff1, bufflen1, MPI_BYTE, MPI_STATUS_IGNORE);
	assert( ierr == MPI_SUCCESS );
#endif
	bwritten += bufflen1 ;

      } else {
	ierr = MPI_File_read_at_all_end(ifh,buff2,MPI_STATUS_IGNORE);
	assert( ierr == MPI_SUCCESS );
	bufflen2 = MIN(fsize-bwritten,bufflen2) ;
	bread    += bufflen2 ;

	bufflen1 = MIN(fsize-bread   ,bufflen1) ;
	if ( bufflen1 > 0 )
	  ierr = MPI_File_read_at_all_begin(ifh, bread, buff1, bufflen1, MPI_BYTE);
	assert( ierr == MPI_SUCCESS );

#ifndef DO_NOT_WRITE
	ierr = MPI_File_write_at(ofh, bwritten, buff2, bufflen2, MPI_BYTE, MPI_STATUS_IGNORE);
	assert( ierr == MPI_SUCCESS );
#endif
	bwritten += bufflen2 ;

      }
      count++;
    }

    /*
     * End
     */

#ifndef DO_NOT_WRITE
    ierr = MPI_File_close(&ofh);
    assert( ierr == MPI_SUCCESS );
#endif
    ierr = MPI_File_close(&ifh);
    assert( ierr == MPI_SUCCESS );
    ierr = MPI_Info_free(&ifinfo);
    assert( ierr == MPI_SUCCESS );

    timing = MPI_Wtime() - timing;

    ierr = rename(ofilename, ofilename2);
    if (ierr == -1)
      fprintf(stderr,"#Error: could not rename file %s (rank=%i) \n", ofilename, rank );

    fprintf(stderr,"#copy size, timing: %s %g (s) %s/s (rank=%i).\n",
	    readable_fs( (double)fsize , ctmp),
	    timing,
	    readable_fs( (double)(fsize/timing),ctmp2),
	    rank);

    /* free memory */
    free(buff1);
    free(buff2);

  }

  /* exit */

  MPI_Finalize();
}


glost_filter.sh
#!/bin/bash

filter_line_forced="sed -r -e 's/#executed[^:]*:[[:space:]]*//g' "
filter_line=${filter_line_forced}

function PrintHelpAndExit(){
    cat <<EOF
glost_filter.sh \- filter the logs of glost_launch(1).

SYNOPSIS

   glost_filter.sh [ACTION] [PRINTCONTROL] [-R <command_file>] <logfile>[...]

   ACTION      is   [-a|-h|-n|-s <status>]
   PRINTCONTROL is   [-A|-X][-C]

DESCRIPTION

   glost_filter.sh  filter the logs of glost_launch(1).

   By default, it will simply extracts the list of commands executed.
   (equivalent to option -aX)

OPTIONS

  ACTION

  -h    
        print this help and exit.
  -H    
        list all commands and highlight the commands with non-zeros status.
        This is equivalent to -s '[^0][[:digit:]]*[[:space:]].*|$' -C . 
  -l    
        list all commands executed regarless of their status.
  -n    
        only show commands with non zeros status.
        This is equivalent to -s '[^0][[:digit:]]*[[:space:]].*' . 
  -s <status>   
        only show command with status <status>

  PRINTCONTROL

  -A    show all informations the executed command (process, time, status).
  -X    only show the executed command
  -C    colorize the matches.

  OTHER
  
  -R <command_file>
        show the remaining command from <command_file>.
        This only compatible with PRINTCONTROL "-X".

        Warning : 
        By default, this option avoids that a same command is launched multiple times.
        To do that, it will removes all occurences of <command> in <command_file>.

        If you want to avoid this, simply add a specific comment at the end of line :

        before                          after 
        <same_command>                  <same_command> #1
        <same_command>                  <same_command> #2
        
        You can see an example (on "sleep 0.1" and "sleep 0.2" ): 
        $cd examples 
        $../glost_filter.sh -R test_filter.list1 test_filter.log1 

EXAMPLES 

   cd examples
   # list all command executed
   ../glost_filter.sh test_filter.log1 
   # list all command executed , highlight non-zero status
   ../glost_filter.sh -H test_filter.log1 
   # list all command executed with non zeros status
   ../glost_filter.sh -n test_filter.log1 
   # list all command executed with non zeros status, see their status
   ../glost_filter.sh -nA test_filter.log1 
   # list all commands undone regardless of their status
   ../glost_filter.sh -R test_filter.list1 test_filter.log1 
   # list all commands undone or which exited with non zeros status
   ../glost_filter.sh -s 0 -R test_filter.list1 test_filter.log1 

SEE ALSO
glost_launch(1) # cd man ; groff -man -Tascii glost_launch.1 | less
EOF
exit
}

# show_remaining <string> <logfile>[...]
function show_bystatus()
{
    string=$1
    shift
    grep -hE '^#[^:]*:.*$' $* |\
    grep ${GCOLOR} -E "^.*status[[:space:]]${string}.*$"
}

# show_remaining <CMDFILE> <string> <logfile>[...]
function show_remaining(){
    TMPFILE=`mktemp`
    CMDFILE=$1
    string=$2
    shift 2
    show_bystatus ${string} $* | eval ${Pline} > ${TMPFILE}
    grep -vxF -f ${TMPFILE} ${CMDFILE}
    rm ${TMPFILE}
}



# MAIN
Pline_whole="cat"
Pline_command="sed -r -e 's/#executed[^:]*:[[:space:]]*//g' "
Pline=${Pline_command}

CMDFILE=""
string='.*'
while getopts hHlns:AXCR: opt
 do
    case $opt in
	# ACTION
	h) PrintHelpAndExit ;;
	l) string='.*' ;;
	s) string="${OPTARG}" ;;
	H) string='[^0][[:digit:]]*[[:space:]].*|$' 
	   GCOLOR="--color=always" ;;
	n) string='[^0][[:digit:]]*[[:space:]].*'   ;;
	# PRINTCONTROL
	A) Pline=${Pline_whole} ;;
	X) Pline=${Pline_command} ;;
	C) GCOLOR="--color=always";;
	# OTHER
	R) CMDFILE="${OPTARG}" ;;
    esac
done
shift $((OPTIND-1))

[[ "$#" == "0" ]] && PrintHelpAndExit

if [[ "${CMDFILE}" == "" ]] ; then
    show_bystatus ${string} $* | eval ${Pline}
else
    [[ "${Pline}" != "${Pline_command}" ]] && echo "Bad option, -R forces -X" && exit 1
    show_remaining $CMDFILE ${string} $* ;
fi



glost_launch.c
/*
   GLOST - Greedy Launcher Of Small Tasks.
   Copyright (C) 2014  Bruno FROGE <bruno.froge@cea.fr>

   Author: Vincent DUCROT     <vincent.ducrot.tgcc@cea.fr>
   Author: Yohan LEE-TIN-YIEN <yohan.lee-tin-yien.tgcc@cea.fr> 

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA

*/


#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <signal.h>
#include <sys/wait.h>
#include <errno.h>
#include <limits.h>


#define GLOST_VERSION_MAJOR 0
#define GLOST_VERSION_MINOR 3
#define GLOST_VERSION_PATCH 1
static const char flavor[]="-tgcc";

#define GLOST_STRLEN 4096
#define GLOST_MANAGER 0
#define TAG_AVAIL 1
#define TAG_TSK   2

/* core functions */
void read_and_exec(char *filename, double user_tremain, size_t len);
void read_and_send(int workers, char *filename, double user_tremain, size_t len);
void recv_and_exec(int rank, size_t len);
static void read_next_command(FILE *fl,char *tsk, double user_tremain, size_t len);
static void exec_command(char *tsk, int rank);

/* options handling */
int read_options(int argc, char *argv[],double *user_tremain,size_t *maxstrlen);
void print_options(char *argv0);

/* signal handling */
void set_sigaction();
static void set_glost_no_new_task(int signum);

#ifdef __HAVE_LIBCCC_USER__
#include "ccc_user.h"
#else
static int ccc_tremain(double *tremain)
{
  tremain[0] = 0. ;
  return EXIT_FAILURE;
}
#endif

/********
 * main *
 ********/

static volatile int glost_no_new_task=0;


int main(int argc,char **argv)
{
  int rank,size;
  int tmp;
  double user_tremain = -1.0 ;
  size_t maxstrlen = GLOST_STRLEN ;

  /* init */
  MPI_Init(&argc,&argv);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  MPI_Comm_size(MPI_COMM_WORLD,&size);  

  tmp=0;
  set_sigaction();

  /* read options */
  if (rank == GLOST_MANAGER)
    tmp=read_options(argc,argv,&user_tremain,&maxstrlen);
  
  MPI_Bcast(&tmp,1,MPI_INT,0,MPI_COMM_WORLD);
  if (tmp != 0){
    MPI_Finalize();
    exit(0);
  }
  MPI_Bcast(&maxstrlen,1,MPI_UINT64_T,0,MPI_COMM_WORLD);

  if (rank == GLOST_MANAGER) {
    printf("Manager is process %d, nb workers: %d\n", GLOST_MANAGER,size-1);
  }
  
  /* launch */
  if (size == 1)
    read_and_exec(argv[optind],user_tremain,maxstrlen);
  else if(size != 1 && rank == GLOST_MANAGER)
    read_and_send(size-1,argv[optind],user_tremain,maxstrlen);
  else
    recv_and_exec(rank,maxstrlen);

  /* exit */
  MPI_Finalize();
}

/****************** 
 * core functions *
 ******************/

void read_and_exec(char *filename, double user_tremain, size_t len )
{
  FILE *fl;
  char tsk[len];

  fl = fopen(filename,"r");
  if (fl == NULL){
    fprintf(stderr,"#WARNING could not open %s: skipping it.\n", filename);
    return;
  }
  do {
    read_next_command(fl,tsk,user_tremain,len);
    if(!strlen(tsk))
      break;
    exec_command(tsk,GLOST_MANAGER);
  } while (strlen(tsk));
  fclose(fl);
}

void read_and_send(int workers,char *filename, double user_tremain, size_t len)
{
  int i, num;
  MPI_Status status;
  FILE *fl;
  char tsk[len];

  fl = fopen(filename,"r");
  if (fl == NULL){
    fprintf(stderr,"#WARNING could not open %s: skipping it.\n", filename);
    /* the rest is handled in read_next_command() to properly kill the processes */
  }

  i=0;
  do {
    read_next_command(fl,tsk,user_tremain,len);
    
    MPI_Recv(&num,1,MPI_INT,MPI_ANY_SOURCE,
	     TAG_AVAIL,MPI_COMM_WORLD,&status);
    MPI_Send(tsk,len,MPI_CHAR,num,
	     TAG_TSK,MPI_COMM_WORLD);
    if (!strlen(tsk)) /* terminate command sent */
      i++;
  } while (i < workers);
  fclose(fl);
}

void recv_and_exec(int rank,size_t len)
{
  char tsk[len];
  MPI_Status status;  

  do {
    MPI_Send(&rank,1,MPI_INT,GLOST_MANAGER,
	     TAG_AVAIL,MPI_COMM_WORLD); 
    MPI_Recv(tsk,len,MPI_CHAR,GLOST_MANAGER,
	     TAG_TSK,MPI_COMM_WORLD,&status); 
    if(!strlen(tsk)) /* terminate command */
      break;
    exec_command(tsk,rank);
  } while (strlen(tsk));
}

static void read_next_command(FILE *fl,char *tsk, double user_tremain, size_t len)
{
  int r;
  double tremain;

  tsk[0]='\0'; 

  r = ccc_tremain(&tremain); 	
  r = (!r && tremain < user_tremain);
  if (r) 
    fprintf(stderr,"#time remaining = %g < %g : do not launch new tasks.\n",
	    tremain,user_tremain);

  if (glost_no_new_task) {
      fprintf(stderr,"#signal received: no new task will be submitted\n");
      r = 1;
  }

  if (r) return;

  /* if(!fgets(tsk,GLOST_MAXLEN,fl)) tsk[0]='\0'; */
  /* if(tsk[strlen(tsk)-1]=='\n') tsk[strlen(tsk)-1]='\0'; */

  do {
    if(!fgets(tsk,len,fl)){
      tsk[0] = '\0';
      break;
    } 
    r = strlen(tsk);
  } while (r < 2);

  if(tsk[r-1]=='\n') tsk[r-1]='\0';

}

static void exec_command(char *tsk, int rank)
{
  double t;
  int r;

  t = MPI_Wtime();  
  r = system(tsk);
  t = MPI_Wtime() - t;
  if ( WIFEXITED(r) ) 
    {
      r = WEXITSTATUS(r); 
      fprintf(stderr,"#executed by process\t%d in %gs\twith status\t%d : %s\n",
	      rank,t,r,tsk);
    }
  else { 
    if( WIFSIGNALED(r) ){
      r = WTERMSIG(r); 
      fprintf(stderr,"#executed by process\t%d in %gs\tkilled by signal %d: %s\n",
	      rank,t,r,tsk);
    }
  }
}


/******************** 
 * option functions *
 ********************/


int read_options(int argc, char *argv[],double *user_tremain,size_t *maxstrlen)
{
  int r;
  int longoptind;
  user_tremain[0] = -1.;
  maxstrlen[0] = GLOST_STRLEN ;

  static struct option longopts[]=
    {
      {"help", no_argument, NULL, 'h' },
      {"version", no_argument, NULL, 'v' },
      {"time_remaining",required_argument,NULL,'R'},
      {"maxline_length",required_argument,NULL,'l'},
      {0,0,0,0}
    };

  while ( (r = getopt_long(argc, argv,"hvl:R:",longopts,&longoptind)) != -1 )
    {
      if (r == 0){ 		/* long option */
	fprintf(stderr,"r %c\n",longopts[longoptind].val);	
	r = longopts[longoptind].val ;
      }
	
      switch(r)
	{
	case 'h': 
	  print_options(argv[0]);
	  return 1;
	  break;
	case 'v': 
	  fprintf(stdout,"%d.%d.%d%s\n",GLOST_VERSION_MAJOR,GLOST_VERSION_MINOR,GLOST_VERSION_PATCH,flavor);	
	  return 1;
	  break;
	case 'l':
	  errno = 0;    /* To distinguish success/failure after call */
	  *maxstrlen = strtoul(optarg,NULL,10);
	  if (errno != 0) {
	    perror("strtol");
	    return 1;
	  }
	  break;
	case 'R': 
	  *user_tremain = strtod(optarg,NULL);
#ifndef   __HAVE_LIBCCC_USER__
	  fprintf(stderr,"WARNING option -R %g ignored: not compiled with libccc_user\n",*user_tremain);
#endif
	  break;

	case '?':
	  fprintf(stderr,"WARNING option %c ignored: unknown option\n",optopt);	
	  return 1;
	  break;

	default:
	  return 1;
	  break;
	}
    }

  if (optind == argc) {
    fprintf(stderr,"ERROR: no input file. Exiting...\n");
    return 1;
  }

  return 0;
}

void print_options(char *argv0){

  /* NAME */
  printf("\nNAME\n\n");
  printf("glost_launch - Greedy launcher Of Small Tasks.\n");

  /* SYNOPSIS */
  printf("\nSYNOPSIS\n\n");
  printf("Usage: [mpirun [mpi_option]...] %s [OPTION] TASK_FILE... \n", argv0);  

  printf("\n"
	 "Where \'mpirun\', is your MPI launcher,\n"
	 "'mpi_option\' is your MPI option.\n"
	 "'TASK_FILE\' is the file containing the independent tasks,\n"
	 "one task per line.\n"
	 );

  /* OPTION */
  printf("\nOPTION\n\n");

  printf(""
    "-h, --help          display this help and exit\n"
    "-v, --version       display version information and exit\n"
    "-l,--maxline_length[=]size \n"
    "   specify the maximum number of character per line in the input file \n"
    "    (default is %i , maximal value is %lu)\n"
    "-R,--time_remaining[=]time \n"
    "   stop submit new tasks if the time remaining for the job is inferior to \'time\' \n"
    "   It is only available if you have compiled it with libccc_user. \n"
    "   (-D __HAVE_LIBCCC_USER__) \n",GLOST_STRLEN,ULONG_MAX );

  /* DESCRIPTION */

  printf("\nDESCRIPTION\n\n");

  printf(""
  "\'%s\' try to execute simultaneously \"number of MPI processes -1\" lines\n"
  "of \'TASK_FILE\' in parrallel.\n"
  "\n"
  "This will help you launching a maximum of independent tasks,\n"
  "with variable execution time under the limitations of your cluster batch sheduler.\n",argv0);

  printf(""
  "\n"
  "  - Algorithm -\n"
  "\n"
  "One invocation of %s starts an MPI job.\n"
  "Process 0 is called manager, while the others, workers.\n"
  "\n"
  "The manager reads the next task, waits for \n"
  "a free worker, and sends the task to the free worker. \n"
  "\n"
  "Meanwhile, each worker says that he is free to the manager, \n"
  "waits for the task, and executes it. \n"
  "\n"
  "If %s is launched in sequential, the manager  \n"
  "reads and executes himself each task.  \n"
  "\n"
  ,argv0,argv0);

  printf(
  "\n"
  "  - Outputs -\n"
  "\n"
  "Each task is executed with \'system(3)\', timed, and\n"
  "its status is logged on stderr, using the following format : \n"
  "\n"
  "  #executed by process <rank> in <time>s with status <status> : <task> \n"
  "\n");

  printf(
  "\n"
  "  - Ending %s during execution. -\n"
  "\n"
  "Instead of using the traditional SIGQUIT, use SIGUSR1 to kill \n"
  "%s properly (to make %s stop launching new commands).\n"
  "\n"
  "Unfortunately, we cannot use the traditional SIGQUIT \n"
  "because OpenMPI and its derivates filter the signal.  \n"
  "For an example, see example/test_sigusr1.sh  \n"
  ,argv0,argv0,argv0);

  printf(
  "\n"
  "  - Process Environment -\n"
  "\n"
  "The task launched inherits their environment from the user's shell.\n"
  "\n"
  "Thus user may use exported variables in the \'TASK_FILE\' file.\n"
  "We do advise using them for paths.\n"
  );

  printf(
  "\n"
  "  - Process Environment -\n"
  "\n"
  "If %s is killed, it should kill every worker.   \n"
  "\n"
  "If a task fails, its returned code is showed in <status> in the outputs.  \n"
  "The user may use glost_filter.sh to extracts from a logged stderr,  \n"
  "which task was run, which one are remainings, etc. \n",argv0);

  /* See also */
  printf("\nSEE ALSO\n\n");
  printf("glost_filter.sh -h\n");

}


/******************* 
 * signal handling *
 *******************/

void set_sigaction()
{
  struct sigaction new_action;

  new_action.sa_handler = set_glost_no_new_task ;
  sigemptyset(&new_action.sa_mask);
  new_action.sa_flags = SA_RESETHAND;

  if (sigaction(SIGUSR1,&new_action,NULL) < 0)
    fprintf(stderr,"#WARNING sigaction failed: no SIGQUIT support\n");
}

static void set_glost_no_new_task(int signum)
{
  glost_no_new_task=1;
}

